### PR TITLE
arbitrary frequency grid for DFT functions in C interface

### PR DESCRIPTION
**Pull request by oskooi**
Progress towards #1070 (still needs Python interface). Closes #1141.

Adds support for arbitrary frequency grids for DFT functions. There are now two interfaces to all DFT functions: (1) the original equally-spaced grid specified by (`freq_min`, `freq_max`,`Nfreq`) and (2) the new unequally-spaced grid specified by `std::vector<double>`.

ToDo: Python interface and tutorial example to be added in a separate PR.

### CODE REVIEW

**Comment by oskooi** [comment]
SWIG typemaps for the `freq` array are added. However, 2 of the 40 tests in the Python suite are failing (`dft_energy.py` and `geom.py`). Not sure why calling `sim.add_energy` in `dft_energy.py` is causing a segmentation fault.

**Comment by stevengj** [comment]
As I commented when we discussed #1141, we should probably be using `std::vector<double> freqs` for every argument where you are passing `double *freqs, size_t Nfreqs`.

Then if you do `flux_object.freqs` in Python it will return an array of frequencies.

File: python/simulation.py [hunk 1]
```diff
@@ -2967,7 +2975,7 @@ def scale_flux_fields(s, flux):
 
 
 def get_flux_freqs(f):
-    return np.linspace(f.freq_min, f.freq_min + f.dfreq * f.Nfreq, num=f.Nfreq, endpoint=False).tolist()
+    return np.linspace(f.freq[0], f.freq[-1], num=f.Nfreq, endpoint=True).tolist()
```
Comment (stevengj) [inline python/simulation.py]: This will just become `return f.freqs`

File: python/meep.i [hunk 2]
```diff
@@ -1346,29 +1366,19 @@ void _get_eigenmode(meep::fields *f, double omega_src, meep::direction d, const
...
     %pythoncode %{
         def freqs(self):
             import math
             import numpy as np
-            start = self.omega_min / (2 * math.pi)
-            stop = start + (self.domega / (2 * math.pi)) * self.Nomega
-            return np.linspace(start, stop, num=self.Nomega, endpoint=False).tolist()
+            start = self.omega[0] / (2 * math.pi)
+            stop = self.omega[-1] / (2 * math.pi)
+            return np.linspace(start, stop, num=self.Nomega, endpoint=True).tolist()
```
Comment (stevengj) [inline python/meep.i]: After the change to `std::vector`, just `return (self.omega / (2*math.pi)).tolist()`

File: python/meep.i [hunk 3]
```diff
@@ -1346,29 +1366,19 @@ void _get_eigenmode(meep::fields *f, double omega_src, meep::direction d, const
...
 %extend meep::dft_ldos {
 
-    double get_omega_min() {
-        return $self->omega_min;
-    }
-    double get_domega() {
-        return $self->domega;
-    }
     int get_Nomega() {
         return $self->Nomega;
```
Comment (stevengj) [inline python/meep.i]: This will be `return $self->freqs.size()`

Comment (stevengj) [reply python/meep.i]: We might not need `get_Nomega` anymore, since this was only used internally

**Comment by stevengj** [comment]
The `dft_ldos` accessor functions were added by @ChristopherHogan in #581.  It's a bit weird that we have these for `dft_ldos` but not for flux objects etcetera?   Presumably this code will simplify a *lot* (and maybe eliminate the SWIG 4 incompatibility #965) if we just have a `std::vector` object for the frequency array.

Along the way, we probably want to change the `dft_ldos` object to use `freqs` rather than `omega` to be consistent with the rest of the code.

**Comment by stevengj** [comment]
Note that for each function there should be two interfaces: the old interface that takes `(freq_min, freq_max, Nfreqs)` and the new interface that takes `freqs`.   However, the implementation of the old interface should just be a one-liner that calls the new interface by constructing an array from `(freq_min, freq_max, Nfreqs)` .   You'll probably want to add a helper function `meep::linspace(freq_min, freq_max, Nfreqs)` that creates the `freqs`.

**Comment by oskooi** [comment]
`double *freq` is changed to `std::vector<double> freq` and all C tests are passing. However, it seems that SWIG is *not* automatically generating typemaps for `std::vector<double> freq` in Python since `flux.freq.tolist()` is not being recognized as indicated by many of the Python tests which use `get_flux_freqs` are failing with:

```
AttributeError: 'DoubleVector' object has no attribute 'tolist'
```

**Comment by oskooi** [comment]
After removing the `.tolist` suffix to `freq` as well as some other minor fixes, *all* tests are now passing. I also fixed a warning bug in `python/tests/geom.py` caused by [`python/simulation.py:725`](https://github.com/NanoComp/meep/blob/master/python/simulation.py#L725). If everything looks good, the final thing to add is a C test.

File: python/simulation.py [hunk 4]
```diff
@@ -2104,9 +2107,9 @@ def get_eigenmode_coefficients(self, flux, bands, eig_parity=mp.NO_PARITY, eig_v
             direction = flux.normal_direction
 
         num_bands = len(bands)
-        coeffs = np.zeros(2 * num_bands * flux.Nfreq, dtype=np.complex128)
-        vgrp = np.zeros(num_bands * flux.Nfreq)
-        cscale = np.zeros(num_bands * flux.Nfreq)
+        coeffs = np.zeros(2 * num_bands * flux.freq.size(), dtype=np.complex128)
+        vgrp = np.zeros(num_bands * flux.freq.size())
+        cscale = np.zeros(num_bands * flux.freq.size())
```
Comment (stevengj) [inline python/simulation.py]: Does every call to `flux.freq.size()` first create a numpy array out of `flux.freq`?

Comment (stevengj) [reply python/simulation.py]: I guess not, because `.size()` is not a NumPy array method.

File: python/simulation.py [hunk 5]
```diff
@@ -2967,7 +2970,7 @@ def scale_flux_fields(s, flux):
 
 
 def get_flux_freqs(f):
-    return np.linspace(f.freq_min, f.freq_min + f.dfreq * f.Nfreq, num=f.Nfreq, endpoint=False).tolist()
+    return f.freq
```
Comment (stevengj) [inline python/simulation.py]: What type does this return?

Comment (stevengj) [reply python/simulation.py]: (i.e. what is `type(f.freq)`?)

Comment (oskooi) [reply python/simulation.py]:
```
>>> type(freqs)
<class 'meep.DoubleVector'>
```

Comment (stevengj) [reply python/simulation.py]: Probably want to convert it to a Python list if it is a Meep array.

File: src/dft.cpp [hunk 6]
```diff
@@ -338,17 +367,23 @@ void load_dft_hdf5(dft_chunk *dft_chunks, component c, h5file *file, const char
 dft_flux::dft_flux(const component cE_, const component cH_, dft_chunk *E_, dft_chunk *H_,
                    double fmin, double fmax, int Nf, const volume &where_,
                    direction normal_direction_, bool use_symmetry_)
-    : Nfreq(Nf), E(E_), H(H_), cE(cE_), cH(cH_), where(where_), normal_direction(normal_direction_),
+    : E(E_), H(H_), cE(cE_), cH(cH_), where(where_), normal_direction(normal_direction_),
+      use_symmetry(use_symmetry_) {
```
Comment (stevengj) [inline src/dft.cpp]: Can't you use the initializer for this?   i.e. have `freq(linspace(fmin, fmax, Nf)) { }`

File: python/simulation.py (hunk 5 above)
Comment (stevengj) [reply python/simulation.py]: Try
```py
[freqs[i] for i in range(freqs.size())]
```
to make a Python list.

File: src/near2far.cpp [hunk 7]
```diff
@@ -46,8 +59,9 @@ dft_near2far::dft_near2far(dft_chunk *F_, double fmin, double fmax, int Nf, doub
 }
 
 dft_near2far::dft_near2far(const dft_near2far &f)
-    : freq_min(f.freq_min), dfreq(f.dfreq), Nfreq(f.Nfreq), F(f.F), eps(f.eps), mu(f.mu),
-      where(f.where) {
+    : F(f.F), eps(f.eps), mu(f.mu), where(f.where) {
```
Comment (stevengj) [inline src/near2far.cpp]: You can just do `freq(f.freq)` as part of the initializers and it should make a copy — no need to copy the elements one by 1.

File: src/stress.cpp [hunk 8]
```diff
@@ -27,20 +27,27 @@ namespace meep {
...
+  freq = meep::linspace(fmin, fmax, Nf);
+  offdiag1 = offdiag1_;
+  offdiag2 = offdiag2_;
+  diag = diag_;
+  // where = new volume(where_.get_min_corner(), where_.get_max_corner());
+}
+
+dft_force::dft_force(dft_chunk *offdiag1_, dft_chunk *offdiag2_, dft_chunk *diag_,
+                     const std::vector<double> freq_, const volume &where_)
+    : where(where_) {
```
Comment (stevengj) [inline src/stress.cpp]: As above: just add `freq(freq_)` to the initializers list

File: src/stress.cpp [hunk 9]
```diff
@@ -27,20 +27,27 @@ namespace meep {
...
   diag = diag_;
   // where = new volume(where_.get_min_corner(), where_.get_max_corner());
 }
 
 dft_force::dft_force(const dft_force &f) : where(f.where) {
-  freq_min = f.freq_min;
-  Nfreq = f.Nfreq;
-  dfreq = f.dfreq;
+  for (int i = 0; i < f.freq.size(); ++i)
+    freq.push_back(f.freq[i]);
```
Comment (stevengj) [inline src/stress.cpp]: `freq(f.freq)` in the initializers list

Comment (stevengj) [reply src/stress.cpp]: Or just `freq = f.freq`

**Pull request merged**